shiny/ui/_theme_brand.py

```diff
@@ -0,0 +1,528 @@
+              @if $contrast  < 4.5 {
+                @warn "The contrast ratio of #{$contrast} between the brand's foreground color (#{inspect($foreground)}) and background color (#{inspect($background)}) is very low. Consider picking colors with higher contrast for better readability.";
+              }
+
+              $white: if($lum_fg > $lum_bg, $foreground, $background);
+              $black: if($lum_fg <= $lum_bg, $foreground, $background);
+
+              // If the brand foreground/background are close enough to black/white, we
+              // use those values. Otherwise, we'll mix the white/black from the brand
+              // fg/bg with actual white and black to get something much closer.
+              $result: (
+                "white": if(contrast-ratio($white, white) <= 1.15, $white, mix($white, white, 20%)),
+                "black": if(contrast-ratio($black, black) <= 1.15, $black, mix($black, black, 20%)),
+              );
+
+              @return $result;
+            }
+            """
+        )
+        self.add_defaults(
+            """
+            $enable-brand-grays: true !default;
+            // Ensure these variables exist so that we can set them inside of @if context
+            // They can still be overwritten by the user, even with !default;
+            $white: null !default;
+            $black: null !default;
+            $gray-100: null !default;
+            $gray-200: null !default;
+            $gray-300: null !default;
+            $gray-400: null !default;
+            $gray-500: null !default;
+            $gray-600: null !default;
+            $gray-700: null !default;
+            $gray-800: null !default;
+            $gray-900: null !default;
+
+            @if $enable-brand-grays {
+              @if variable-exists(brand--foreground) and variable-exists(brand--background) {
+                $brand-white-black: brand-choose-white-black($brand--foreground, $brand--background);
+                @if $white == null {
+                  $brand-white: map-get($brand-white-black, "white");
+                  @if $brand-white != null {
+                    $white: $brand-white;
+                  }
+                }
+                @if $black == null {
+                  $brand-black: map-get($brand-white-black, "black");
+                  @if $brand-black != null {
+                    $black: $brand-black;
+                  }
+                }
+              }
+              @if $white != null and $black != null {
+                $gray-100: mix($white, $black, 90%);
+                $gray-200: mix($white, $black, 80%);
+                $gray-300: mix($white, $black, 70%);
+                $gray-400: mix($white, $black, 60%);
+                $gray-500: mix($white, $black, 50%);
+                $gray-600: mix($white, $black, 40%);
+                $gray-700: mix($white, $black, 30%);
+                $gray-800: mix($white, $black, 20%);
+                $gray-900: mix($white, $black, 10%);
```

Comment:
I think all these need default flags (otherwise they can't be overridden)?

Reply:
If the user has set them, they won't be `null`. https://sass-lang.com/playground/#eJxNizEKwCAQBPt7xYaktUirCPmKoJLA4YFoGvHvibFJt8zO7PvmQ9SIIlje5SoXQzRhqsx/elwRJdcAlzyGAmun1AiYjZQzZEOdSGeR8j1K3Y5r0FjbkDr1B3jNJAM=

Reply:
That said, it won't hurt to add the `!default` flag.